components/common/src/test/java/com/hotels/styx/configstore/ConfigStoreTest.java

```diff
@@ -0,0 +1,162 @@
+/*
+  Copyright (C) 2013-2018 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.configstore;
+
+import com.hotels.styx.configstore.ConfigStore.ConfigEntry;
+import com.hotels.styx.support.Latch;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+import rx.Observable;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.hotels.styx.support.matchers.IsOptional.isValue;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.is;
+
+public class ConfigStoreTest {
+    private ConfigStore configStore;
+
+    @BeforeMethod
+    public void setUp() {
+        configStore = new ConfigStore();
+    }
+
+    @Test
+    public void getsWhatWasSet() {
+        configStore.set("foo", "bar");
+        assertThat(configStore.get("foo", String.class), isValue("bar"));
+    }
+
+    @Test
+    public void listenersReceiveUpdatesWhenValuesChange() {
+        Latch sync = new Latch(1);
+        AtomicReference<String> update = new AtomicReference<>();
+
+        configStore.watch("foo", String.class)
+                .subscribe(value -> {
+                    update.set(value);
+                    sync.countDown();
+                });
+
+        configStore.set("foo", "bar");
+        sync.await(1, SECONDS);
+        assertThat(update.get(), is("bar"));
+        assertThat(configStore.get("foo", String.class), isValue("bar"));
+    }
+
+    // If this test fails it will cause a deadlock, resulting in a latch timeout
+    @Test
+    public void listensOnSeparateThread() {
+        Latch unlockedByTestThread = new Latch(1);
+        Latch unlockedBySubscribeThread = new Latch(1);
+
+        configStore.watch("foo", String.class)
+                .subscribe(value -> {
+                    unlockedByTestThread.await(1, SECONDS);
+                    unlockedBySubscribeThread.countDown();
+                });
+
+        configStore.set("foo", "bar");
+        unlockedByTestThread.countDown();
+        unlockedBySubscribeThread.await(2, SECONDS);
```

Comment:
CyclicBarrier seems more appropriate for this use case.

Reply:
I don't know what difference that would make.

Reply:
It would be using only one synchronising primitive instead of two. That cuts the complexity by 50% ;-).

Reply:
I think this is conceptually much simpler:

```
    public void listensOnSeparateThread2() {
        final CyclicBarrier barrier = new CyclicBarrier(2);

        configStore.watch("foo", String.class)
                .subscribe(value -> {
                        await(barrier);
                });

        configStore.set("foo", "bar");
        await(barrier);
    }
```

Reply:
I think this will work. I'll try it.

Reply:
This is not sufficient to test the functionality. The ordering matters. If we use `CyclicBarrier` then a `watch` method that uses the same thread as the unit test would still pass, since both awaits will get called.

To prove that a separate thread is used, the `watch` must be able to wait for the test thread. That said, it probably only needs the `unlockedByTestThread` latch, not both.

Reply:
OK.